[Deprecation] Properly deprecate CentOS7 as CI build image/Supported OS and switch to Almalinux8

###### 20240425 We have since moved from Rockylinux8 to Almalinux8
###### 20240430 We will switch back to CentOS7 for another release in 2.14.0

**[Deprecation] Properly deprecate CentOS7 as CI build image/Supported OS and switch to RockyLinux8**

Back in #3743, we already plan to migrate all CentOS7 images to RockyLinux8 as part of the NodeJS18(#1563) / Python3.9(#3351) Upgrade.

However, it was pushed back by the k-NN Team as moving from CentOS7 to RockyLinux8 would change the minimal glibc version from 2.17 to 2.28 during lib build, causing people with servers containing glibc lower than 2.28 failed to run k-NN, as the plugin will crash upon api call.
* https://github.com/opensearch-project/opensearch-build/pull/4002

CentOS7 will end its support by 2024/06/30 (https://endoflife.date/centos), meaning we need to migrate to RockyLinux8 by then, especially for the only build image for OpenSearch. We also need to properly announce it on the website, as our compatibility chart is clearly mentioning the support for it here: https://github.com/opensearch-project/documentation-website/blob/054dc56d5dcef66576cd4b15bef0ae5efa4f8f77/_install-and-configure/install-opensearch/index.md?plain=1#L26

We need to have a plan and decide when do we announce:
1. Discontinue CentOS7 support on Release
2. Discontinue CentOS7 support on CI Build
  2.1. Which in turn, k-NN drop CentOS7 support on its lib as well


Thanks.

------

* More related: https://github.com/opensearch-project/opensearch-build/issues/4687

------

PRs:
* https://github.com/opensearch-project/opensearch-build/pull/4464
* https://github.com/opensearch-project/opensearch-build/pull/4645
* https://github.com/opensearch-project/opensearch-build/pull/4647
* https://github.com/opensearch-project/k-NN/pull/1634
* https://github.com/opensearch-project/opensearch-build/pull/4663
* https://github.com/opensearch-project/k-NN/pull/1664
* https://github.com/opensearch-project/opensearch-build/pull/4667
* https://github.com/opensearch-project/k-NN/pull/1674
* https://github.com/opensearch-project/opensearch-build/pull/4714
* https://github.com/opensearch-project/documentation-website/pull/7287
* https://github.com/opensearch-project/opensearch-build/pull/4744
* https://github.com/opensearch-project/opensearch-build/pull/4760
* https://github.com/opensearch-project/opensearch-build/pull/4809
* #4820 
* https://github.com/opensearch-project/opensearch-build/pull/4903

Apparently gcc10 still have bug on arm64, go back to gcc11 for now.

```
[  2%] Linking CXX shared library release/libopensearchknn_common.so
[  2%] Built target opensearchknn_common
[  2%] Building CXX object external/nmslib/similarity_search/src/CMakeFiles/NonMetricSpaceLib.dir/distcomp_bregman.cc.o
In file included from /home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc:17:
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc: In function 'T similarity::ItakuraSaitoPrecompSIMD(const T*, const T*, size_t) [with T = float; size_t = long unsigned int]':
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/include/utils.h:53:53: note: '#pragma message: /home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc(83) : WARNING: ItakuraSaitoPrecompSIMD<float>: SSE2 is not available, defaulting to pure C++ implementation!'
   53 | #   define WARN(exp) (FILE_LINE_LINK "WARNING: " exp)
      |                                                     ^
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc:83:17: note: in expansion of macro 'WARN'
   83 | #pragma message WARN("ItakuraSaitoPrecompSIMD<float>: SSE2 is not available, defaulting to pure C++ implementation!")
      |                 ^~~~
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc: In function 'T similarity::KLPrecompSIMD(const T*, const T*, size_t) [with T = float; size_t = long unsigned int]':
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/include/utils.h:53:53: note: '#pragma message: /home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc(217) : WARNING: KLPrecompSIMD<float>: SSE2 is not available, defaulting to pure C++ implementation!'
   53 | #   define WARN(exp) (FILE_LINE_LINK "WARNING: " exp)
      |                                                     ^
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc:217:17: note: in expansion of macro 'WARN'
  217 | #pragma message WARN("KLPrecompSIMD<float>: SSE2 is not available, defaulting to pure C++ implementation!")
      |                 ^~~~
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc: In function 'T similarity::KLGeneralPrecompSIMD(const T*, const T*, size_t) [with T = float; size_t = long unsigned int]':
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/include/utils.h:53:53: note: '#pragma message: /home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc(329) : WARNING: KLGeneralPrecompSIMD<float>: SSE2 is not available, defaulting to pure C++ implementation!'
   53 | #   define WARN(exp) (FILE_LINE_LINK "WARNING: " exp)
      |                                                     ^
/home/ci-runner/k-NN/jni/external/nmslib/similarity_search/src/distcomp_bregman.cc:329:17: note: in expansion of macro 'WARN'
  329 | #pragma message WARN("KLGeneralPrecompSIMD<float>: SSE2 is not available, defaulting to pure C++ implementation!")
      |                 ^~~~
Assembler messages:
Error: unknown architectural extension `ssbs'
Error: unrecognized option -march=armv8.2-a+crypto+fp16+rcpc+dotprod+ssbs
make[2]: *** [external/nmslib/similarity_search/src/CMakeFiles/NonMetricSpaceLib.dir/build.make:76: external/nmslib/similarity_search/src/CMakeFiles/NonMetricSpaceLib.dir/distcomp_bregman.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:457: external/nmslib/similarity_search/src/CMakeFiles/NonMetricSpaceLib.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

I tried above option1 which installed all the necessary libs from source, I found the error only happened in [jni_test part](https://github.com/opensearch-project/k-NN/blob/main/jni/CMakeLists.txt#L129-L182) with:
```
[ 83%] Linking CXX executable bin/jni_test
external/faiss/faiss/libfaiss.a(IndexAdditiveQuantizerFastScan.cpp.o): In function `faiss::IndexAdditiveQuantizerFastScan::search(long, float const*, long, float*, long*, faiss::SearchParameters const*) const':
IndexAdditiveQuantizerFastScan.cpp:(.text+0xc88): undefined reference to `void faiss::IndexFastScan::search_dispatch_implem<true>(long, float const*, long, float*, long*, faiss::NormTableScaler const*) const'
release/libopensearchknn_faiss.so: undefined reference to `_gfortran_concat_string'
collect2: error: ld returned 1 exit status
make[2]: *** [bin/jni_test] Error 1
``` 

After I removed the jni_test part and run `./gradlew integTest`, all integration test can be passed:
```
> Task :integTest
May 22, 2024 5:09:20 PM sun.util.locale.provider.LocaleProviderAdapter <clinit>
WARNING: COMPAT locale provider will be removed in a future release

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.4/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 26m 7s
19 actionable tasks: 19 executed
[ci-runner@a26d5cd0a0e9 k-NN]$ arch
aarch64
```

So maybe we can bypass the jni_test part for AL2 arm64 env build.

Ref for the dockerfile diff: https://github.com/opensearch-project/opensearch-build/compare/main...junqiu-lei:opensearch-build:knn-build-1

--- a/manifests/3.0.0/opensearch-3.0.0.yml
+++ b/manifests/3.0.0/opensearch-3.0.0.yml
@@ -5,7 +5,7 @@ build:
   version: 3.0.0
 ci:
   image:
-    name: opensearchstaging/ci-runner:ci-runner-centos7-opensearch-build-v3
+    name: opensearchstaging/ci-runner:ci-runner-al2-opensearch-build-v1
     args: -e JAVA_HOME=/opt/java/openjdk-21
 components:
   - name: OpenSearch

